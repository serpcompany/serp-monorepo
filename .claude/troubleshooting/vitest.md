https://vitest.dev/guide/common-errors.html