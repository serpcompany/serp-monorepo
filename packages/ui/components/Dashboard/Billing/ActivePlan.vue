<script setup lang="ts">
  import { useDateFormat } from '@vueuse/core'

  interface BillingPlan {
    id: string
    name: string
    description?: string
    status?: string
    currentPeriodEnd?: Date
    currentPeriodStart?: Date
    amount: number
    interval: string
    priceId: string
    cancelAt?: Date
  }

  interface Props {
    currentPlan: BillingPlan
  }

  const props = defineProps<Props>()

  function formatPrice(price?: number): string {
    if (!price)
      return '$0'
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 0,
    }).format(price / 100)
  }

  function getStatusColor(status?: string): string {
    switch (status) {
    case 'active':
      return 'success'
    case 'trialing':
      return 'primary'
    case 'canceled':
    case 'incomplete_expired':
    case 'unpaid':
      return 'error'
    case 'past_due':
    case 'incomplete':
      return 'warning'
    default:
      return 'neutral'
    }
  }

  function getSubscriptionMessage(plan: BillingPlan): string {
    if (plan.cancelAt) {
      return 'Cancels on'
    }
    if (plan.status === 'trialing') {
      return 'Trial ends on'
    }
    return 'Renews on'
  }

  const formatedDate = computed(() => {
    const date = props.currentPlan.cancelAt || props.currentPlan.currentPeriodEnd
    return useDateFormat(date, 'MMM DD, YYYY')
  })
</script>

<template>
  <UPageCard>
    <template #title>
      You are on
      <span class="font-bold text-highlighted">{{ currentPlan.name }}</span>
      plan
    </template>
    <div class="flex flex-wrap items-center gap-3">
      <div class="flex items-center gap-2">
        <span class="text-xl font-semibold">
          {{ formatPrice(currentPlan.amount) }}
        </span>
        <span class="text-sm text-muted">
          every {{ currentPlan.interval }}
        </span>
      </div>
      <UBadge
        :label="currentPlan.status"
        :color="getStatusColor(currentPlan.status)"
        variant="subtle"
        class="capitalize"
      />
      <span class="text-sm text-muted">
        {{ getSubscriptionMessage(currentPlan) }}
        {{ formatedDate }}
      </span>
    </div>
  </UPageCard>
</template>
