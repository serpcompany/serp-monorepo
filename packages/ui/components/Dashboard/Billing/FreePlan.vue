<template>
  <UPageCard>
    <template #title>
      You are on the
      <span class="font-bold text-highlighted">Free</span>
      plan
    </template>
    <template #description>
      Upgrade to a paid plan to unlock more features and higher usage
      limits.
    </template>
  </UPageCard>
</template>
