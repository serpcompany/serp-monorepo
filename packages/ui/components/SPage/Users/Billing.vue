<script setup lang="ts">
  const { loggedIn } = useUserSession();
  if (!loggedIn.value) {
    navigateTo('/');
  }

  const router = useRouter();

  const isLoading = ref(false);

  async function createStripeBillingPortalSession() {
    isLoading.value = true;
    try {
      const response = await $fetch('/api/stripe/portal', {
        method: 'GET'
      });
      if (response) {
        window.open(response, '_blank');
      }
    } catch (error) {
      console.error('Error creating billing portal session:', error);
    } finally {
      isLoading.value = false;
    }
  }

  useSeoMeta({
    title: 'Manage Billing'
  });
</script>

<template>
  <UPage>
    <UMain>
      <UContainer class="mx-auto mt-96 max-w-3xl">
        <UCard class="my-8">
          <template #header>
            <div class="flex items-center justify-between">
              <h2 class="text-xl font-semibold">Subscription Management</h2>
              <UBadge color="primary" variant="soft">Active</UBadge>
            </div>
          </template>

          <div class="px-4 text-center">
            <p class="mb-6 text-gray-500 dark:text-gray-400">
              Manage your subscription, view billing history, or update payment
              methods.
            </p>

            <UButton
              :loading="isLoading"
              :disabled="isLoading"
              label="Manage Billing"
              color="primary"
              size="lg"
              icon="i-lucide-credit-card"
              variant="solid"
              class="mx-auto"
              @click="createStripeBillingPortalSession"
            />
          </div>
        </UCard>
      </UContainer>
    </UMain>
  </UPage>
</template>
