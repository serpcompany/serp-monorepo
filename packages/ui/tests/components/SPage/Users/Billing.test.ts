import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import { describe, expect, it } from 'vitest';
import { ref } from 'vue';
import ComponentRender from '../../../componentRender';
import Billing from '../../../../components/SPage/Users/Billing.vue';

// Declare a mutable session_ variable.
let session_: { loggedIn: unknown; user: unknown; clear: unknown };
// Set the global mock once.
(globalThis as unknown).useUserSession = () => session_;

mockNuxtImport('useSeoMeta', () => () => {});

describe('SPage/Users/Billing Snapshot', () => {
  const scenarios: [
    string,
    { session: { loggedIn: unknown; user: unknown; clear: unknown } }
  ][] = [
    [
      'when user is logged in',
      {
        session: {
          loggedIn: ref(true),
          user: ref({
            name: 'Test User',
            email: 'test@test.com',
            image: 'https://example.com/test.jpg'
          }),
          clear: vi.fn()
        }
      }
    ],
    [
      'when user is not logged in',
      {
        session: {
          loggedIn: ref(false),
          user: ref(null),
          clear: vi.fn()
        }
      }
    ]
  ];

  it.each(scenarios)(
    'renders %s correctly',
    async (desc: string, { session }) => {
      // Update the global session for this scenario.
      session_ = session;

      const html = await ComponentRender(
        `SPage/Users/Billing ${desc}`,
        {},
        Billing
      );
      expect(html).toMatchSnapshot();
    }
  );
});
