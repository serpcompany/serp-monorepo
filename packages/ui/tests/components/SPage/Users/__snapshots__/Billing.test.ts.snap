// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SPage/Users/Billing Snapshot > renders when user is logged in correctly 1`] = `
"<div class="flex flex-col lg:grid lg:grid-cols-10 lg:gap-10">
  <!--v-if-->
  <div class="lg:col-span-10">
    <main class="min-h-[calc(100vh-var(--ui-header-height))]">
      <div class="px-4 sm:px-6 lg:px-8 mx-auto mt-96 max-w-3xl">
        <div class="rounded-[calc(var(--ui-radius)*2)] bg-(--ui-bg) ring ring-(--ui-border) divide-y divide-(--ui-border) my-8">
          <div class="p-4 sm:px-6">
            <div class="flex items-center justify-between">
              <h2 class="text-xl font-semibold">Subscription Management</h2><span class="font-medium inline-flex items-center text-xs px-2 py-1 gap-1 rounded-[calc(var(--ui-radius)*1.5)] bg-(--ui-primary)/10 text-(--ui-primary)"><!--v-if-->Active<!--v-if--></span>
            </div>
          </div>
          <div class="p-4 sm:p-6">
            <div class="px-4 text-center">
              <p class="mb-6 text-gray-500 dark:text-gray-400"> Manage your subscription, view billing history, or update payment methods. </p><button type="button" class="rounded-none font-medium inline-flex items-center focus:outline-hidden disabled:cursor-not-allowed aria-disabled:cursor-not-allowed disabled:opacity-75 aria-disabled:opacity-75 transition-colors px-3 py-2 text-sm gap-2 text-(--ui-bg) bg-(--ui-primary) hover:bg-(--ui-primary)/75 disabled:bg-(--ui-primary) aria-disabled:bg-(--ui-primary) focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-(--ui-primary) mx-auto"><span class="iconify i-lucide:credit-card shrink-0 size-5" aria-hidden="true"></span><span class="truncate">Manage Billing</span>
                <!--v-if-->
              </button>
            </div>
          </div>
          <!--v-if-->
        </div>
      </div>
    </main>
  </div>
  <!--v-if-->
</div>"
`;

exports[`SPage/Users/Billing Snapshot > renders when user is not logged in correctly 1`] = `
"<div class="flex flex-col lg:grid lg:grid-cols-10 lg:gap-10">
  <!--v-if-->
  <div class="lg:col-span-10">
    <main class="min-h-[calc(100vh-var(--ui-header-height))]">
      <div class="px-4 sm:px-6 lg:px-8 mx-auto mt-96 max-w-3xl">
        <div class="rounded-[calc(var(--ui-radius)*2)] bg-(--ui-bg) ring ring-(--ui-border) divide-y divide-(--ui-border) my-8">
          <div class="p-4 sm:px-6">
            <div class="flex items-center justify-between">
              <h2 class="text-xl font-semibold">Subscription Management</h2><span class="font-medium inline-flex items-center text-xs px-2 py-1 gap-1 rounded-[calc(var(--ui-radius)*1.5)] bg-(--ui-primary)/10 text-(--ui-primary)"><!--v-if-->Active<!--v-if--></span>
            </div>
          </div>
          <div class="p-4 sm:p-6">
            <div class="px-4 text-center">
              <p class="mb-6 text-gray-500 dark:text-gray-400"> Manage your subscription, view billing history, or update payment methods. </p><button type="button" class="rounded-none font-medium inline-flex items-center focus:outline-hidden disabled:cursor-not-allowed aria-disabled:cursor-not-allowed disabled:opacity-75 aria-disabled:opacity-75 transition-colors px-3 py-2 text-sm gap-2 text-(--ui-bg) bg-(--ui-primary) hover:bg-(--ui-primary)/75 disabled:bg-(--ui-primary) aria-disabled:bg-(--ui-primary) focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-(--ui-primary) mx-auto"><span class="iconify i-lucide:credit-card shrink-0 size-5" aria-hidden="true"></span><span class="truncate">Manage Billing</span>
                <!--v-if-->
              </button>
            </div>
          </div>
          <!--v-if-->
        </div>
      </div>
    </main>
  </div>
  <!--v-if-->
</div>"
`;
